### ISSUE TITLE
Add git resource option to not checkout to new branch

### ISSUE THREAD

**Issue by spiffytech** [chef/chef#1984]
The `git` resource checks out code to a new branch ("deploy", by default). In my organization, this doesn't seem to provide any benefit, and actually causes causes quite a bit of confusion. We have various teams and tools that log onto machines expecting to muck with the git checkout, and get confused at this random branch that's not tracking any upstream repo, and doesn't match any project ticket numbers.

Since in my circumstance - and I imagine many others\* - it makes sense to have 'git branch' reflect whatever was placed into the "revision" parameter, I'd like the `git` resource to have a parameter enabling this behavior. I.e., if `revision` is set to `master`, `git branch` should say "master", and the repository should be set to track origin/master, from which it was originally cloned (but still respect `action`). If `revision` is set to `e7f4630`, `git branch` should say "(detached from e7f4630)".

\* I actually can't determine the benefit of the deploy branch in any circumstance, but I'm assuming the person who coded the behavior knew of one.

**Comment by iiro** [comment]
:+1:

**Comment by josegonzalez** [comment]
Yeah it would be nice to have this. My current move is an LWRP that checks out the code and then notifies an execute to switch to back to master/delete the deploy branch. I can *maybe* see the benefit if you're using this under the `application` lwrp, but it's a bit of a drag if you're using chef to manage a workstation.

**Comment by zhimsel** [comment]
I'd like to also say this would be very useful. I'd rather have the checked out branch be the *actual* branch, not 'deploy'.

**Comment by kiaplayer** [comment]
:+1:

**Comment by amaltson** [comment]
:+1:, just had to add a manual `execute` resource to fix this up to the master branch.

**Comment by stevendanna** [comment]
I would need to look into whether this would break the intuition of how this works.  But in omnibus, we basically always use ls-remote to resolve whatever we have to a full SHA.  Then we git fetch and then git checkout that sha.  I think that might work here as well?

**Comment by mintigo-efi** [comment]
Seconded. In fact we immediately do 
```
execute "git checkout" do
	command "git checkout #{node['branch']}"
```
right after `git` recipe. I never understood the reason for deploying into detached head.

**Comment by spiffytech** [comment]
@stevendanna While I think seeing the sha rather than a `deploy` branch is a very minor improvement, I don't think it solves most of the problems prompting this issue. It still leaves the user confusion over what's actually checked out (what branch? what feature? what ticket?), and leaves the repo in a state where you can't pull changes from upstream without manual intervention (pain for users, problem for scripts running on the system).

**Comment by PowerKiKi** [comment]
Same here, being detached on a hardcoded branch is very unexpected.

**Comment by lamont-granquist** [comment]
I somewhat have my doubts that this is actually a breaking change, since it looks like the kind of feature that nobody wants, but someone out there over the past 10 years has probably become bug-dependent upon this behavior.

**Comment by lamont-granquist** [comment]
closed by #9726
